sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DescribeTableExec.scala

```diff
@@ -0,0 +1,96 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalog.v2.{Identifier, TableCatalog}
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.encoders.RowEncoder
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, GenericRowWithSchema}
+import org.apache.spark.sql.catalyst.plans.DescribeTableSchemas
+import org.apache.spark.sql.execution.LeafExecNode
+import org.apache.spark.sql.sources.v2.Table
+
+case class DescribeTableExec(
+    catalog: TableCatalog,
+    ident: Identifier,
+    isExtended: Boolean) extends LeafExecNode {
+
+  import DescribeTableExec._
+
+  override def output: Seq[AttributeReference] = DescribeTableSchemas.DESCRIBE_TABLE_ATTRIBUTES
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    val rows = new ArrayBuffer[InternalRow]()
+    if (catalog.tableExists(ident)) {
+      val table = catalog.loadTable(ident)
+      addSchema(rows, table)
+
+      if (isExtended) {
+        addPartitioning(rows, table)
+        addProperties(rows, table)
+      }
+
+    } else {
+      rows += toCatalystRow(s"Table $ident does not exist.", "", "")
+    }
+    sparkContext.parallelize(rows)
+  }
+
+  private def addSchema(rows: ArrayBuffer[InternalRow], table: Table): Unit = {
+    rows ++= table.schema.map{ column =>
+      toCatalystRow(
+        column.name, column.dataType.simpleString, column.getComment().getOrElse(""))
+    }
+  }
+
+  private def addPartitioning(rows: ArrayBuffer[InternalRow], table: Table): Unit = {
+    rows += EMPTY_ROW
+    rows += toCatalystRow(" Partitioning", "", "")
+    rows += toCatalystRow("--------------", "", "")
+    if (table.partitioning.isEmpty) {
+      rows += toCatalystRow("Not partitioned", "", "")
+    } else {
+      rows ++= table.partitioning.zipWithIndex.map {
+        case (transform, index) => toCatalystRow(s"Part $index", transform.describe(), "")
+      }
+    }
+  }
+
+  private def addProperties(rows: ArrayBuffer[InternalRow], table: Table): Unit = {
+    rows += EMPTY_ROW
+    rows += toCatalystRow(" Table Property", " Value", "")
+    rows += toCatalystRow("----------------", "-------", "")
+    rows ++= table.properties.asScala.toList.sortBy(_._1).map {
+      case (key, value) => toCatalystRow(key, value, "")
+    }
+  }
+}
+
+private object DescribeTableExec {
+  private val ENCODER = RowEncoder(DescribeTableSchemas.DESCRIBE_TABLE_SCHEMA)
+  private val EMPTY_ROW = toCatalystRow("", "", "")
+
+  private def toCatalystRow(strs: String*): InternalRow = {
+    ENCODER.resolveAndBind().toRow(
```

Comment:
the encoder only need to call `resolveAndBind` once

Reply:
I don't necessarily think so, but it could also be how this class is built. I think the encoder's state needs to be reset. When I don't resolveAndBind every time, the tests yield wrong results entirely.

Reply:
Fixed it - we have to copy the rows generated by the encoder since the encoder re-uses the same memory space.